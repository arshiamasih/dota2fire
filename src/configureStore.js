import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import reducers from "./reducer/index";

export default function configureStore(initialState) {
  return createStore(reducers, initialState, applyMiddleware(thunk));
}
