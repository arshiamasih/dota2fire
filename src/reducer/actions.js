import {ADD_WINNERS, GET_TEAMS, CREATE_GAME, CALL_API, RECEIVE_API, GET_PLAYERS, CLOSE_MODAL} from './types'

export const requestAPI = () => ({
  type: CALL_API,
  payload: {
    status: 'pending'
  }
  
})

export const receiveAPI = () => ({
  type: RECEIVE_API,
  payload: {
    status: 'success'
  }
  
})

export const hideModal = () => ({
  type: CLOSE_MODAL,
  payload: false
  
})

export const addWinners = (round, match, winner) => 
({ 
    type: ADD_WINNERS,
    payload: {
      round : winner
    },
    round: round+1,
    match
  })

export const fetchTeamData = (teams) => 
  ({
    type: GET_TEAMS,
    payload: teams
  })


export const fetchTeamPlayers = (players, name) => 
({
  type: GET_PLAYERS,
  payload: {
    team: name, 
    names: players
  },
  modalExpand: true
})  

export const createGame = (n, structure) => 
  ({
    type: CREATE_GAME,
    payload: 
     {
      n,
      num: n*2,
      start: true
    },
    structure
  })

export const getGameNum = (n) => async (dispatch) =>{
  const structure = createStructure(n)
  return dispatch(createGame(n, structure))
}
export const getTeams = (num) => async (dispatch) => {
  //API ALREADY RANKS BY ELO SCORE
  //remove the no name team member
  //add try catch
  try {
    dispatch(requestAPI())
  const response = await fetch('https://api.opendota.com/api/teams')
  const data = await response.json()
  dispatch(receiveAPI())
  const teams = data.slice(0,num);
  console.log('in actions', teams)
  return dispatch(fetchTeamData(teams))
  }
  catch (error) {
    //dispatch server failure
  }
}


export const getPlayers = (players, name) => async (dispatch) => {
  return dispatch(fetchTeamPlayers(players, name))

}

export const closeModal = () => async (dispatch) => {
  return dispatch(hideModal())
}

export const getWinner = (round, match, winner) => async (dispatch) => {
  return dispatch(addWinners(round, match, winner))
}

const createStructure = (n) => {
  const arr = createMatchesHash([], n, Math.log2(n))
  const obj = {0: []}
  for(let i =0; i < arr.length; i++) {
      obj[i+1] = {matches:arr[i]}
  }
  return obj
}

//helper for createStructure
const createMatchesHash = (arr = [], n, i ) => {
  //base case
  if(arr.length === i) return arr
  else {
   n /=2 
   const node = {}
   for(let i = 0; i < n; i++) {
     node[i] = {
       win: false,
       name: null
     }
   }
   arr.push(node) 
   return createMatchesHash(arr, n, i ) 
  }

}





