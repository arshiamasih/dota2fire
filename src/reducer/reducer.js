import { ADD_WINNERS, GET_TEAMS, CREATE_GAME, CALL_API, RECEIVE_API, FAILED_API, GET_PLAYERS, CLOSE_MODAL } from "./types";


const initialState = {
  apiStatus: {
    teams: null,
    players: null
  },
  modalExpand: false,
  teams: [],
  players: {
    team: null,
    names:[]
  },
  winner: {
     0: [], 
    1: 
    {
      matches: {
      0: {
        name: null,
        win: false
        },
      },
    },
  },   
  gameNum: {
    n: 0,
    num: 0,
    start: false
  }
}

const teamReducer=(state = initialState, action) => {
  switch (action.type) {
    case CREATE_GAME: 
    return {
      ...state,
      gameNum: Object.assign({}, action.payload),
      winner: action.structure
    }
    case ADD_WINNERS:
      //dynamic assignment to tree structure    
      return  {
        ...state,
        winner: {
          ...state.winner,  
        [action.round]: {
            ...state.winner[action.round],
            matches: {
            ...state.winner[action.round].matches,
            [action.match]: Object.assign({},{ 
            name: action.payload.round,
           })},
         },
        }
      };
    case GET_TEAMS: 
      return {
        ...state,
        teams: action.payload,
        winner: {
          ...state.winner,
          0: action.payload
        }
      };
    case GET_PLAYERS: 
    return {
      ...state,
      modalExpand: true,
      players: Object.assign({}, {
        ...action.payload
      })
    }
    case CALL_API:
      return {
        ...state,
        apiStatus: {
          ...state.apiStatus,
          [action.call]: action.payload
        }
      } 
    case RECEIVE_API:
      return {
        ...state,
        apiStatus: {
          ...state.apiStatus,
          [action.call]: action.payload
        }
    }
    case FAILED_API:
      return {
        ...state,
        apiStatus: {
          ...state.apiStatus,
          [action.call]: action.payload
        }
    }   
    case CLOSE_MODAL: 
    return {
      ...state,
      modalExpand: action.payload,
    }   
    default:
      return state;  
  }
}
export default teamReducer


