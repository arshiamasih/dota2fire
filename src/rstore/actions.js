import {ADD_WINNERS, GET_TEAMS} from './types'


export const addWinners = () => ({
  type: ADD_WINNERS
})

export const getTeams = () => ({
  type: GET_TEAMS,
  payload: {
    teams: [
      {id: 1,
        name: 'Warriors',
        rank: 1,
        isWinner: false
      },
      {id: 2,
        name:'Vikings',
        rank: 3,
        isWinner: false},    
    ]
  }
})

