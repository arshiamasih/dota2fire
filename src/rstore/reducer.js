import { ADD_WINNERS, GET_TEAMS } from "./types";

const initialState = {
  rounds: {
    winners: []
  },
  teams: {}
}
const teamReducer=(state = initialState, action) => {
  switch (action.type) {
    case ADD_WINNERS:
      return {
        ...state
      };
    case GET_TEAMS:
      return {
        ...state,
        teams: action.payload
      };  
    default:
      return state;  
  }

}

export default teamReducer