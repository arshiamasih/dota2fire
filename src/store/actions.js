import {ADD_WINNERS, GET_TEAMS, CREATE_GAME} from './types'
export const addWinners = (round, match, winner) => {

return { 
    type: ADD_WINNERS,
    payload: {
      round : winner
    },
    round: round+1,
    match
  }
}



export const fetchTeamData = (teams) => 
  ({
    type: GET_TEAMS,
    payload: teams
  })

export const createGame = (n) => 
  ({
    type: CREATE_GAME,
    payload: {
      n,
      num: n*2,
      start: true
    }
  })

export const getGameNum = (n) => async (dispatch) =>{
  return dispatch(createGame(n))
}
export const getTeams = (num) => async (dispatch) => {
  //API ALREADY RANKS BY ELO SCORE
  //remove the no name team member
  const response = await fetch('https://api.opendota.com/api/teams')
  const data = await response.json()
  const teams = data.slice(0,num);
  return dispatch(fetchTeamData(teams))
}

export const getWinner = (round, match, winner) => async (dispatch) => {
  console.log('action wtfff', round, match, winner)
  return dispatch(addWinners(round, match, winner))
}



