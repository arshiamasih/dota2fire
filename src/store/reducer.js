import { ADD_WINNERS, GET_TEAMS } from "./types";

const initialState = {
  teams: []
}
const teamReducer=(state = initialState, action) => {
  switch (action.type) {
    case ADD_WINNERS:
      return {
        ...state
      };
    case GET_TEAMS:
      console.log('test reducer', state)
      return {
        teams: action.payload.teams
      };  
    default:
      return state;  
  }

}



export default teamReducer