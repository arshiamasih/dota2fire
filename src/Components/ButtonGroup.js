import React from 'react';
import { getGameNum, getTeams, getPlayers } from '../reducer/actions'
import { connect } from 'react-redux'


const GameButtonGroup = (props) => {

  const labels = ['Two', 'Four', 'Eight', 'Sixteen']
  const onClick = async (event)=> {
    event.preventDefault()
    const obj = {
      0: 2,
      1: 4,
      2: 8,
      3: 16,
    }
    props.getGameNum(obj[event.target.id])
    props.getTeams(obj[event.target.id])
  }

  return (
    <div className={'button-group'}>
    {props.gameNum.start? <div ><p><code>Start new game:</code></p></div> : <div ><p><code>Select # of teams:</code></p></div>} 
        <div className={'button-group-btn'} >
        {labels.map((label,i) => {
          return <button className={'game-btn'} key={i} id={i} onClick={event => onClick(event)}>{label}</button>
        }) } 
       </div>
    </div>
  );
}

const mapState = (state) => {
  return {
    gameNum: state.teams.gameNum
  }
}

const mapDispatch = (dispatch) => ({
  getGameNum: (n)=>dispatch(getGameNum(n)),
  getTeams: (n)=>dispatch(getTeams(n)),
})

export default connect(mapState, mapDispatch)(GameButtonGroup)
