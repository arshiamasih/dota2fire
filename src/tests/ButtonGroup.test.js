import React from 'react';
import ButtonGroup from '../Components/ButtonGroup';
import Renderer from 'react-test-renderer'
import configureStore from '../configureStore';
import { Provider } from 'react-redux'
import { getGameNum } from '../reducer/actions'

const mockStore = configureStore

describe('<ButtonGroup/> ', () => {
  let store;
  let component;
  beforeEach(() => {
    store = mockStore(
      {
      teams: {
        teams: [{team_id: 2586976,
          rating: 1614.41,
          wins: 547,
          losses: 301,
          last_match_time: 1581265389,
          name: "OG",
          tag: "OG",
          logo_url: "https://steamcdn-a.akamaihd.net/apps/dota2/images/team_logos/2586976.png"
        }],
        gameNum: {
          n: 16,
          num: 32,
          start: false
        },
        getGameNum: ()=> {},
        getTeams: ()=> {}
    }});

    store.dispatch = jest.fn()
    component = Renderer.create(
      <Provider store={store}>
        <ButtonGroup/>
      </Provider>
    );
  });

  it('should match snapshot', () => {
    expect(component.toJSON()).toMatchSnapshot();
  })

  // it('should dispatch an action on button click', () => {
  //   Renderer.act(() => {
  //     component.root[0].findByType('button').props.onClick();
  //   });
  //   expect(store.dispatch).toHaveBeenCalledTimes(1);
  //   expect(store.dispatch).toHaveBeenCalledWith(
  //     getGameNum({ payload: {n: 0, num: 0, start: true} })
  //   );
  // });
})