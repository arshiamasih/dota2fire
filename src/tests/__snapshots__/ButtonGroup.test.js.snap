// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonGroup/>  should match snapshot 1`] = `
<div
  className="button-group"
>
  <div>
    <p>
      <code>
        Select # of teams:
      </code>
    </p>
  </div>
  <div
    className="button-group-btn"
  >
    <button
      className="game-btn"
      id={0}
      onClick={[Function]}
    >
      Two
    </button>
    <button
      className="game-btn"
      id={1}
      onClick={[Function]}
    >
      Four
    </button>
    <button
      className="game-btn"
      id={2}
      onClick={[Function]}
    >
      Eight
    </button>
    <button
      className="game-btn"
      id={3}
      onClick={[Function]}
    >
      Sixteen
    </button>
  </div>
</div>
`;
